src/receivers/kafka/receiver.rs

```diff
@@ -0,0 +1,892 @@
+    pub traces_topic: &'static str,
+    pub metrics_topic: &'static str,
+    pub logs_topic: &'static str,
+    pub format: DeserializationFormat,
+}
+
+impl KafkaReceiver {}
+
+impl KafkaReceiver {
+    pub fn new(
+        config: KafkaReceiverConfig,
+        traces_output: Option<OTLPOutput<Vec<ResourceSpans>>>,
+        metrics_output: Option<OTLPOutput<Vec<ResourceMetrics>>>,
+        logs_output: Option<OTLPOutput<Vec<ResourceLogs>>>,
+    ) -> Result<Self> {
+        // Get the list of topics to subscribe to
+        let topics = config.get_topics();
+
+        if topics.is_empty() {
+            return Err(KafkaReceiverError::ConfigurationError(
+                "No topics configured for subscription. Enable at least one signal type (traces, metrics, or logs) with a corresponding topic".to_string()
+            ));
+        }
+
+        // Build the Kafka client configuration
+        let cc = config.build_client_config();
+        let consumer: StreamConsumer = cc.create().map_err(|e| {
+            KafkaReceiverError::ConfigurationError(format!("Failed to create receiver: {}", e))
+        })?;
+
+        // Subscribe to the configured topics
+        let topic_refs: Vec<&str> = topics.iter().map(|s| s.as_str()).collect();
+        consumer.subscribe(&topic_refs).map_err(|e| {
+            KafkaReceiverError::ConfigurationError(format!("Failed to subscribe to topics: {}", e))
+        })?;
+
+        let traces_topic = config.traces_topic.clone();
+        let traces_topic = traces_topic.unwrap_or("".into()).leak();
+
+        let metrics_topic = config.metrics_topic.clone();
+        let metrics_topic = metrics_topic.unwrap_or("".into()).leak();
+
+        let logs_topic = config.logs_topic.clone();
+        let logs_topic = logs_topic.unwrap_or("".into()).leak();
+
+        Ok(Self {
+            consumer,
+            traces_output,
+            metrics_output,
+            logs_output,
+            traces_topic,
+            metrics_topic,
+            logs_topic,
+            format: config.deserialization_format,
+        })
+    }
+
+    async fn send_to_pipeline<T>(
+        &self,
+        request: Vec<T>,
+        output: &Option<OTLPOutput<Vec<T>>>,
+    ) -> std::result::Result<(), SendError> {
+        if let Some(output) = output {
+            output.send(request).await?;
+        }
+        Ok(())
+    }
+
+    pub(crate) async fn run(
+        &self,
+        receivers_cancel: CancellationToken,
+    ) -> std::result::Result<(), Box<dyn Error + Send + Sync>> {
+        debug!("Starting Kafka receiver");
+
+        // The consumer will automatically start from the position defined by auto.offset.reset
+        // which is set to "earliest" by default in the config
+        let subscription = self.consumer.subscription().unwrap();
+        debug!("Initial subscriptions: {:?}", subscription);
+
+        loop {
+            select! {
+                     record = self.consumer.recv() => {
```

Comment:
Does handling all of the topics/telemetry types in a single receiver select cause a back pressure impact between telemetry types? For example, if logs were going to Clickhouse and traces going to Honeycomb, and Clickhouse were to have back pressure, would traces be impacted due to blocking on `send_to_pipeline`?

That said, I'm not sure if we wouldn't have other choke points, but I was curious whether running a consumer task per topic/type would help in unequal back-pressure scenarios or on multi-core boxes.

Reply:
Also totally fine if changes are planned to be handled in future PRs :+1: